torch/csrc/autograd/variable.cpp

```diff
@@ -221,17 +217,32 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
   torch::autograd::impl::update_cpp_hooks_on_new_gradfn(self, self.grad_fn());
 }
 
-void create_cpp_hook(const at::TensorBase& self) {
-  auto& list = materialize_autograd_meta(self)->cpp_hooks_list_;
-  // NOLINTNEXTLINE(modernize-make-shared)
-  list.reset(new hooks_list());
-  std::unique_ptr<FunctionPreHook> hook_ptr(
-      new CppFunctionPreHook(list, self.output_nr()));
-  clear_hooks(self);
-  add_hook(self, std::make_shared<CppFunctionPreHook>(list, 0));
+void create_cpp_hook(const at::TensorBase& self, bool is_retains_grad_hook) {
   const auto& fn = self.grad_fn();
-  if (fn) {
-    fn->add_pre_hook(std::move(hook_ptr));
+  if (is_retains_grad_hook) {
+    std::shared_ptr<hooks_list>& list =
+        materialize_autograd_meta(self)->retains_grad_hooks_list_;
+    // NOLINTNEXTLINE(modernize-make-shared)
+    list.reset(new hooks_list());
+    std::unique_ptr<FunctionPreHook> hook_ptr{
```

Comment:
nit: everything above this is shared in both branches, so that can be moved outside of the if statement.

Reply:
Yeah this one looks kinda ugly right now with so much repetition, but I'm not sure how to simplify further at the moment - the top branch uses `retains_grad_hooks_list_`, and the second has `cpp_hooks_list_`. It's tricky because references must be initialized when defined and cannot be mutated to point to different things.